### PR TITLE
Automatic Folio-Keycloak and Folio-Kong Git cloning & minor tenant and UI improvements

### PR DESCRIPTION
**Pull request by BKadirkhodjaev**
## Purpose

- Add automatic Git cloning
- Add tenant & deploy UI controls
- Add other housekeeping improvements
- Add automatic sidecar version determination
- Add more flags to primary deployment commands
- Improve file copying (`setup` and `deployUi` commands)
- Move struct creation methods to dedicated models
- Add `config.combine.yaml` configuration to deploy ACQ-only application with UI
- Disable interface validation for `mgr-application` and `mgr-tenant-entitlements`
- Add module version fixation to negate perturbations caused by snapshot updates

## Approach

- Add Git clone of `folio-keycloak` and `folio-kong` 
- Add `internal.HasTenant` method
- Add `internal.DeployUi` method
- Add `GetFuncName` method
- Partially revert & update `deployUi` command
- Remove `getDefaultSidecar()` from setting a sidecar value for ` mgr-*` modules
- Clean `docker-compose.yaml`, `init.sql`, `config.minimal.yaml` and `config.acquisitions.yaml`

### CODE REVIEW

File: eureka-cli/cmd/deployUi.go [hunk 1]
```diff
@@ -46,49 +44,38 @@ var deployUiCmd = &cobra.Command{
...
 		internal.UpdateKeycloakPublicClientParams(deployUiCommand, enableDebug, tenant, masterAccessToken)
 
-		slog.Info(deployUiCommand, "### COPYING PLATFORM COMPLETE UI CONFIGS ###", "")
-		files, err := filepath.Glob("eureka-tpl/*")
-		if err != nil {
-			// Handle error if the pattern doesn't match any files
-			internal.LogErrorPanic(deployUiCommand, fmt.Sprintf("Failed to glob files: %v", err.Error()))
-		}
+		slog.Info(deployUiCommand, internal.GetFuncName(), "### COPYING PLATFORM COMPLETE UI CONFIGS ###")
+		internal.RunCommandFromDir(deployUiCommand, exec.Command("cp", "-R", "-f", "eureka-tpl/*", "."), outputDir)
```
Comment (steveellis) [inline eureka-cli/cmd/deployUi.go:72]: On linux and probably macos too I get the following when this command runs. I think the if/else below was my attempt to fix this.

```2024/11/19 14:27:49 INFO Deploy UI github.com/folio-org/eureka-cli/cmd.DeployUi="### COPYING PLATFORM COMPLETE UI CONFIGS ###"
cp: cannot stat 'eureka-tpl/*': No such file or directory
2024/11/19 14:27:49 ERROR Deploy UI github.com/folio-org/eureka-cli/internal.RunCommand="systemCmd.Run() error"
panic: exit status 1

goroutine 1 [running]:
github.com/folio-org/eureka-cli/internal.RunCommand({0xe16454, 0x9}, 0xc000833ac8?)
        /home/sellis/folio/eureka-setup/eureka-cli/internal/command.go:14 +0x133
github.com/folio-org/eureka-cli/internal.RunCommandFromDir(...)
        /home/sellis/folio/eureka-setup/eureka-cli/internal/command.go:20
github.com/folio-org/eureka-cli/cmd.DeployUi()
        /home/sellis/folio/eureka-setup/eureka-cli/cmd/deployUi.go:72 +0x785
github.com/folio-org/eureka-cli/cmd.DeployApplication()
        /home/sellis/folio/eureka-setup/eureka-cli/cmd/deployApplication.go:41 +0x45
github.com/folio-org/eureka-cli/cmd.init.func10(0xc00051f000?, {0xe13b70?, 0x4?, 0xe13b74?})
        /home/sellis/folio/eureka-setup/eureka-cli/cmd/deployApplication.go:28 +0xf
github.com/spf13/cobra.(*Command).execute(0x16acf20, {0xc000534cf0, 0x3, 0x3})
        /home/sellis/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:989 +0xa91
github.com/spf13/cobra.(*Command).ExecuteC(0x16af1a0)
        /home/sellis/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1117 +0x3ff
github.com/spf13/cobra.(*Command).Execute(...)
        /home/sellis/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1041
github.com/folio-org/eureka-cli/cmd.Execute()
        /home/sellis/folio/eureka-setup/eureka-cli/cmd/root.go:50 +0x1a
main.main()
        /home/sellis/folio/eureka-setup/eureka-cli/main.go:21 +0xf
```

Comment (BKadirkhodjaev) [reply eureka-cli/cmd/deployUi.go:72]: > On linux and probably macos too I get the following when this command runs. I think the if/else below was my attempt to fix this.
> 
> ```
> cp: cannot stat 'eureka-tpl/*': No such file or directory
> 2024/11/19 14:27:49 ERROR Deploy UI github.com/folio-org/eureka-cli/internal.RunCommand="systemCmd.Run() error"
> panic: exit status 1
> 
> goroutine 1 [running]:
> github.com/folio-org/eureka-cli/internal.RunCommand({0xe16454, 0x9}, 0xc000833ac8?)
>         /home/sellis/folio/eureka-setup/eureka-cli/internal/command.go:14 +0x133
> github.com/folio-org/eureka-cli/internal.RunCommandFromDir(...)
>         /home/sellis/folio/eureka-setup/eureka-cli/internal/command.go:20
> github.com/folio-org/eureka-cli/cmd.DeployUi()
>         /home/sellis/folio/eureka-setup/eureka-cli/cmd/deployUi.go:72 +0x785
> github.com/folio-org/eureka-cli/cmd.DeployApplication()
>         /home/sellis/folio/eureka-setup/eureka-cli/cmd/deployApplication.go:41 +0x45
> github.com/folio-org/eureka-cli/cmd.init.func10(0xc00051f000?, {0xe13b70?, 0x4?, 0xe13b74?})
>         /home/sellis/folio/eureka-setup/eureka-cli/cmd/deployApplication.go:28 +0xf
> github.com/spf13/cobra.(*Command).execute(0x16acf20, {0xc000534cf0, 0x3, 0x3})
>         /home/sellis/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:989 +0xa91
> github.com/spf13/cobra.(*Command).ExecuteC(0x16af1a0)
>         /home/sellis/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1117 +0x3ff
> github.com/spf13/cobra.(*Command).Execute(...)
>         /home/sellis/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1041
> github.com/folio-org/eureka-cli/cmd.Execute()
>         /home/sellis/folio/eureka-setup/eureka-cli/cmd/root.go:50 +0x1a
> main.main()
>         /home/sellis/folio/eureka-setup/eureka-cli/main.go:21 +0xf
> ```

Please try with `deploy-ui: false` on your enabled tenants, currently only `snapshot` branch supports a fully-fletched Eureka UI.

Example config:

- https://github.com/folio-org/eureka-setup/blob/0217041590201d3e4849252a5618892e76df58b7/eureka-cli/config.minimal.yaml#L81
Flag usage: 
- https://github.com/folio-org/eureka-setup/blob/0217041590201d3e4849252a5618892e76df58b7/eureka-cli/cmd/deployUi.go#L64

Platform-complete `eureka-tpl` folder contents (i.e. a specialized `stripes.config.js` with Eureka-only properties):

https://github.com/folio-org/platform-complete/tree/snapshot/eureka-tpl

Comment (steveellis) [reply eureka-cli/cmd/deployUi.go:72]: Also I think this directory (I don't know what it does) always exists. Although I am testing with `-c ./config.minimal.yaml` which uses `snapshot` where the directory does exist. I'm thinking that the expansion `/*` is being treated literally by `exec.Command`.

Comment (BKadirkhodjaev) [reply eureka-cli/cmd/deployUi.go:72]: > Also I don't think this directory (I don't know what it does) always exists. Although I am testing with `-c ./config.minimal.yaml` which uses `snapshot` where the directory does exist. I'm thinking that the expansion `/*` is being treated literally by `exec.Command`.

Got it, I'll rewrite this bit with a more platform agnostic approach

File: eureka-cli/cmd/deployUi.go [hunk 2]
```diff
@@ -46,49 +44,38 @@ var deployUiCmd = &cobra.Command{
...
-			internal.RunCommandFromDir(deployUiCommand, exec.Command("cp", args...), outputDir)
-
-			if stripesBranch == defaultStripesBranch {
-				slog.Info(deployUiCommand, "### PREPARING PLATFORM COMPLETE UI CONFIGS ###", "")
-				internal.PrepareStripesConfigJs(deployUiCommand, outputDir, tenant)
-				internal.PreparePackageJson(deployUiCommand, outputDir, tenant)
-			}
-		} else {
-			slog.Info("No files matched in eureka-tpl/*. Not copying.")
-		}
```
Comment (steveellis) [inline eureka-cli/cmd/deployUi.go:89]: This does the equivalent of `/*` and also doesn't fail if the directory `eureka-tpl` doesn't exist or is empty. Please see comment above.

Comment (BKadirkhodjaev) [reply eureka-cli/cmd/deployUi.go:89]: > This does the equivalent of `/*` and also doesn't fail if the directory `eureka-tpl` doesn't exist or is empty. Please see comment above.

I found it silently avoiding copying all together and building the Okapi-based UI, I'll retest it again tomorrow

Comment (steveellis) [reply eureka-cli/cmd/deployUi.go:89]: If this code works on Windows, I think we have something cross-platform. See what you think.

**Review by SerhiiNosko — APPROVED** [review]

**Review by okolawole-ebsco — APPROVED** [review]

**Pull request merged**